include/grpc++/impl/codegen/completion_queue.h

```diff
@@ -79,12 +82,28 @@ class ServerContext;
 
 extern CoreCodegenInterface* g_core_codegen_interface;
 
+/// EXPERIMENTAL
+/// Creates a Thread Local cache to store the first event
+/// On this completion queue queued from this thread.  Once
+/// initialized, it must be flushed on the same thread.
+class CompletionQueueTLSCache : private GrpcLibraryCodegen {
```

Comment:
Maybe make this a private member class of CompletionQueue (public names are forever)... If we need to we can promote later